drivers/watchdog/virtio_wdt.c

```diff
@@ -0,0 +1,240 @@
+	int ret;
+	u8 *buf;
+
+	if (!vi->wdd_register_done)
+		return -ENODEV;
+
+	buf = kzalloc(sizeof(u8), GFP_KERNEL);
+	if (!vi->busy) {
+		vi->busy = true;
+		reinit_completion(&vi->have_data);
+		register_buffer(vi, (void *)buf, sizeof(u64));
+	}
+
+	ret = wait_for_completion_killable(&vi->have_data);
+	if (ret < 0)
+		goto err;
+
+	if (*buf != 1) {
+		dev_err(&vdev->dev,
+				"Host did not acknowledge buffer correctly");
+		ret = -EINVAL;
+	}
+
+err:
+	vi->busy = false;
+	kfree(buf);
+
+	return ret;
+}
+
+static int virtio_watchdog_start(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog started");
+
+	return 0;
+}
+
+static int virtio_watchdog_stop(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog stop request ignored");
+
+	return 0;
+}
+
+static struct watchdog_info vw_info = {
+	.identity = "virtio-watchdog",
+	.options = WDIOF_KEEPALIVEPING,
+};
+
+static const struct watchdog_ops vw_ops = {
+	.owner = THIS_MODULE,
+	.start = virtio_watchdog_start,
+	.stop = virtio_watchdog_stop,
+	.ping = virtio_watchdog_ping,
+};
+
+static int probe_common(struct virtio_device *vdev)
+{
+	int err, index;
+	struct virtio_watchdog_info *vi = NULL;
+
+	vi = kzalloc(sizeof(struct virtio_watchdog_info), GFP_KERNEL);
+	if (!vi)
+		return -ENOMEM;
+
+	vi->index = index =
+		ida_simple_get(&watchdog_index_ida, 0, 0, GFP_KERNEL);
+	if (index < 0) {
+		err = index;
+		goto err_ida;
+	}
+	sprintf(vi->name, "virtio_watchdog.%d", index);
+	init_completion(&vi->have_data);
+
+	vdev->priv = vi;
+	vi->vdev = vdev;
+
+	vi->vq =
+		virtio_find_single_vq(vdev, virtio_watchdog_recv_done, "input");
+	if (IS_ERR(vi->vq)) {
+		err = PTR_ERR(vi->vq);
+		goto err_find;
+	}
+
+	vi->wdd.info = &vw_info;
+	vi->wdd.ops = &vw_ops;
+	vi->wdd.min_timeout = VW_HEARTBEAT_DEFAULT;
+	vi->wdd.max_timeout = VW_HEARTBEAT_DEFAULT;
+	vi->wdd.timeout = VW_HEARTBEAT_DEFAULT;
+
+	err = watchdog_register_device(&vi->wdd);
+	if (err != 0)
+		goto err_find;
+	vi->wdd_register_done = true;
+
+	return 0;
+
+err_find:
+	ida_simple_remove(&watchdog_index_ida, index);
+err_ida:
+	kfree(vi);
+	return err;
+}
+
+static void remove_common(struct virtio_device *vdev)
+{
+	struct virtio_watchdog_info *vi = vdev->priv;
+
+	if (vi->busy) {
+		wait_for_completion(&vi->have_data);
+		vi->data_avail = 0;
+		complete(&vi->have_data);
+		vi->busy = false;
+	}
+	vdev->config->reset(vdev);
+	if (vi->wdd_register_done) {
+		watchdog_unregister_device(&vi->wdd);
+		vi->wdd_register_done = false;
+	}
+	vdev->config->del_vqs(vdev);
+	ida_simple_remove(&watchdog_index_ida, vi->index);
+	kfree(vi);
+}
+
+static int virtio_watchdog_probe(struct virtio_device *vdev)
+{
+	return probe_common(vdev);
+}
+
+static void virtio_watchdog_remove(struct virtio_device *vdev)
+{
+	remove_common(vdev);
```

Comment:
Not sure this probe_common indirection layer is needed. We could have the code defined here directly, and have `virtio_watchdog_freeze()` call into `virtio_watchdog_remove()` directly.